flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java

```diff
@@ -756,6 +756,20 @@ public void cancel() {
 			// the discovery loop may currently be sleeping in-between
 			// consecutive discoveries; interrupt to shutdown faster
 			discoveryLoopThread.interrupt();
+		} else {
+			// when partition discovery is disabled,
+			// we should still call the close method.
+			// otherwise we may have connection leak,
+			// if open method throws an exception after partitionDiscoverer is constructed.
+			// In this case, run method won't be executed
+			// and partitionDiscoverer.close() won't be called.
+			if (partitionDiscoverer != null) {
+				try {
+					partitionDiscoverer.close();
+				} catch (Exception e) {
+					LOG.error("failed to close partitionDiscoverer", e);
+				}
+			}
```

Comment:
I think we should not close the `partitionDiscoverer` here. Instead we should close it in the `open` method if no longer needed and in the `run` method when leaving the `kafkaFetcher` thread.